pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java

```diff
@@ -218,38 +218,47 @@ protected CompletableFuture<Void> internalDeleteNamespaceAsync(boolean force) {
-                            if (!pulsar().getBrokerService().isSystemTopic(TopicName.get(topic))) {
-                                hasNonSystemTopic = true;
-                                break;
-                            }
+                    ArrayList<String> copyAllPartitionTopics = new ArrayList<>();
+                    boolean hasNonSystemTopic = false;
+                    List<String> allSystemTopics = new ArrayList<>();
+                    List<String> allPartitionedSystemTopics = new ArrayList<>();
+                    for (String topic : allTopics) {
+                        if (!pulsar().getBrokerService().isSystemTopic(TopicName.get(topic))) {
+                            hasNonSystemTopic = true;
+                            copyAllTopics.add(topic);
+                        } else {
+                            allSystemTopics.add(topic);
                         }
-                        if (!hasNonSystemTopic) {
-                            for (String topic : allPartitionedTopics) {
-                                if (!pulsar().getBrokerService().isSystemTopic(TopicName.get(topic))) {
-                                    hasNonSystemTopic = true;
-                                    break;
-                                }
-                            }
+                    }
+                    for (String topic : allPartitionedTopics) {
+                        if (!pulsar().getBrokerService().isSystemTopic(TopicName.get(topic))) {
+                            hasNonSystemTopic = true;
+                            copyAllPartitionTopics.add(topic);
+                        } else {
+                            allPartitionedSystemTopics.add(topic);
                         }
-
+                    }
+                    if (!force) {
                         if (hasNonSystemTopic) {
                             throw new RestException(Status.CONFLICT, "Cannot delete non empty namespace");
                         }
                     }
                     return namespaceResources().setPoliciesAsync(namespaceName, old -> {
                         old.deleted = true;
                         return  old;
-                    }).thenCompose(__ -> {
-                        return internalDeleteTopicsAsync(allTopics);
-                    }).thenCompose(__ -> {
-                        return internalDeletePartitionedTopicsAsync(allPartitionedTopics);
+                    }).thenCompose(ignore -> {
+                        return internalDeleteTopicsAsync(copyAllTopics);
+                    }).thenCompose(ignore -> {
+                        return internalDeletePartitionedTopicsAsync(copyAllPartitionTopics);
+                    }).thenCompose(ignore -> {
+                        return internalDeleteTopicsAsync(allSystemTopics);
+                    }).thenCompose(ignore__ -> {
+                        return internalDeletePartitionedTopicsAsync(allPartitionedSystemTopics);
```

Comment:
We should add a separate test: Delete namespaces that contain both system topic and non-system topic, preferably with topics ordered like this `[__change_event, non-system-topic]`

Reply:
In fact, deleting the namespace in TransactionTest is just like this, there are systemtopic[`__change_event`, `__transaction_buffer_snapshot`] and normal topic[`test`, `testDeleteNamespaceAfterUsedTransaction`] in the namespace.
But you are right, we can add `assertTrue` to make it clear.